PyAlbany compilation error

It looks like PyAlbany failed to compile last night due to a missing include:

```
/nightlyCDash/repos/Albany/PyAlbany/src/PyAlbany_Stokhos.hpp:10:10: fatal error: Stokhos_KL_ExponentialRandomField.hpp: No such file or directory
```
https://sems-cdash-son.sandia.gov/cdash/viewBuildError.php?buildid=33107

My guess is the file included either got removed or renamed in Trilinos.  Do we really need the include?  Would be great if you could have a look @kliegeois .

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -386,6 +386,17 @@ ELSE()
   SET(ALBANY_STK_REBALANCE FALSE)
 ENDIF()
 
+# set optional dependency on Stokhos, for KL expansions
+# MESSAGE("-- Looking for Stokhos:")
+LIST(FIND Trilinos_PACKAGE_LIST Stokhos Stokhos_List_ID)
+IF (Stokhos_List_ID GREATER -1)
+  SET(ALBANY_STOKHOS TRUE)
+  MESSAGE("-- Looking for Stokhos ...                       Found.")
+ELSE()
+  MESSAGE("-- Looking for Stokhos ...                   NOT Found.")
+  SET(ALBANY_STOKHOS FALSE)
+ENDIF()
+
 # set optional dependency on Belos's Epetra subpackage
 IF (EXISTS "${Trilinos_INCLUDE_DIRS}/BelosEpetraAdapter.hpp")
   MESSAGE("-- Looking for BelosEpetra ...                   Found.")

--- a/PyAlbany/src/PyAlbany_Stokhos.cpp
+++ b/PyAlbany/src/PyAlbany_Stokhos.cpp
@@ -3,6 +3,7 @@
 using namespace PyAlbany;
 
 KLExpansion::KLExpansion(int _ndim) : ndim(_ndim) {
+#ifdef ALBANY_STOKHOS    
     domain_upper = Teuchos::Array<double>(ndim, 1.0);
     domain_lower = Teuchos::Array<double>(ndim, 0.0);
     correlation_lengths = Teuchos::Array<double>(ndim, 1.0);
@@ -12,22 +13,32 @@ KLExpansion::KLExpansion(int _ndim) : ndim(_ndim) {
     tol = 1e-10;
     max_it = 100;
     num_KL_terms = 1;
+#endif
 }
 
 void KLExpansion::setUpperBound(int i, double _domain_upper_i) {
+#ifdef ALBANY_STOKHOS 
     domain_upper[i] = _domain_upper_i;
+#endif
 }
 void KLExpansion::setLowerBound(int i, double _domain_lower_i) {
+#ifdef ALBANY_STOKHOS 
     domain_lower[i] = _domain_lower_i;
+#endif
 }
 void KLExpansion::setCorrelationLength(int i, double _correlation_length_i) {
+#ifdef ALBANY_STOKHOS 
     correlation_lengths[i] = _correlation_length_i;
+#endif
 }
 void KLExpansion::setNumberOfKLTerm(int i, double _num_KL_terms_i) {
+#ifdef ALBANY_STOKHOS 
     num_KL_per_dim[i] = _num_KL_terms_i;
+#endif
 }
 
 void KLExpansion::createModes() {
+#ifdef ALBANY_STOKHOS 
     Teuchos::ParameterList solverParams;
     solverParams.set("Number of KL Terms", num_KL_terms);
     solverParams.set("Mean", RST::zero());
@@ -42,9 +53,11 @@ void KLExpansion::createModes() {
     solverParams.set("Number of KL Terms per dimension", num_KL_per_dim);
 
     randomField = RandomFieldType(solverParams);
+#endif
 }
 
 void KLExpansion::getModes(double* phi, int n_nodes, int n_modes, double* x, int n_nodes_x) {
+#ifdef ALBANY_STOKHOS 
     Kokkos::View<double *, Kokkos::LayoutLeft, PyTrilinosVector::node_type::device_type> weights("w", n_modes);
     for (std::size_t i_mode = 0; i_mode<n_modes; ++ i_mode) {
         weights(i_mode) = 0;
@@ -58,9 +71,11 @@ void KLExpansion::getModes(double* phi, int n_nodes, int n_modes, double* x, int
             weights(i_mode) = 0.;
         }
     }
+#endif
 }
 
 void KLExpansion::getModes(double* phi, int n_nodes, int n_modes, double* x, int n_nodes_x, double* y, int n_nodes_y) {
+#ifdef ALBANY_STOKHOS 
     Kokkos::View<double *, Kokkos::LayoutLeft, PyTrilinosVector::node_type::device_type> weights("w", n_modes);
     for (std::size_t i_mode = 0; i_mode<n_modes; ++ i_mode) {
         weights(i_mode) = 0;
@@ -74,9 +89,11 @@ void KLExpansion::getModes(double* phi, int n_nodes, int n_modes, double* x, int
             weights(i_mode) = 0.;
         }
     }
+#endif
 }
 
 void KLExpansion::getModes(double* phi, int n_nodes, int n_modes, double* x, int n_nodes_x, double* y, int n_nodes_y, double* z, int n_nodes_z) {
+#ifdef ALBANY_STOKHOS 
     Kokkos::View<double *, Kokkos::LayoutLeft, PyTrilinosVector::node_type::device_type> weights("w", n_modes);
     for (std::size_t i_mode = 0; i_mode<n_modes; ++ i_mode) {
         weights(i_mode) = 0;
@@ -90,10 +107,13 @@ void KLExpansion::getModes(double* phi, int n_nodes, int n_modes, double* x, int
             weights(i_mode) = 0.;
         }
     }
+#endif
 }
 
 void KLExpansion::getEigenValues(double* eigenvalues, int n_modes) {
+#ifdef ALBANY_STOKHOS 
     for (std::size_t i_mode = 0; i_mode<n_modes; ++ i_mode) {
         eigenvalues[i_mode] = randomField.eigenvalue(i_mode);
     }
+#endif
 }

--- a/PyAlbany/src/PyAlbany_Stokhos.hpp
+++ b/PyAlbany/src/PyAlbany_Stokhos.hpp
@@ -7,7 +7,12 @@
 #ifndef PYALBANY_STOKHOS_H
 #define PYALBANY_STOKHOS_H
 
+// Get all Albany configuration macros
+#include "Albany_config.h"
+
+#ifdef ALBANY_STOKHOS
 #include "Stokhos_KL_ExponentialRandomField.hpp"
+#endif
 #include "Albany_PyAlbanyTypes.hpp"
 
 namespace PyAlbany
@@ -16,7 +21,9 @@ namespace PyAlbany
     {
     private:
 #ifndef SWIG
+#ifdef ALBANY_STOKHOS
         typedef Stokhos::KL::ExponentialRandomField<double> RandomFieldType;
+#endif
         typedef Teuchos::ScalarTraits<double> RST;
 
         const int ndim;
@@ -24,8 +31,10 @@ namespace PyAlbany
         Teuchos::Array<int> num_KL_per_dim;
         double eps, tol;
         int max_it, num_KL_terms;
+#ifdef ALBANY_STOKHOS
         RandomFieldType randomField;
-#endif        
+#endif
+#endif
 
     public:
         KLExpansion(int _ndim = 2);

--- a/src/Albany_config.h.in
+++ b/src/Albany_config.h.in
@@ -21,6 +21,7 @@
 #cmakedefine ALBANY_STK_REBALANCE
 #cmakedefine ALBANY_TEKO
 #cmakedefine ALBANY_ZOLTAN
+#cmakedefine ALBANY_STOKHOS
 
 // ========================== General Albany build and capabilities options ============================== //
 

